Implement new specs for push & pop in asm parser

grjte opened the pull request:
This PR addresses the second stage of #84 as well as the new mem ops specified in #102 

- changes push handling for decimal/hexadecimal constants to allow multiple values in the format `push.a.b.c...` up to 16 values
- adds new hex parsing to allow multiple hex values to be specified without separators, as long as the total number of specified bytes is a multiple of 8 (e.g. `push.0x0000000000001234000000000000abcd`)
- adds `push.mem`/`push.mem.a` and `pop.mem`/`pop.mem.a` to push and pop single elements from/to memory
- stubs out `push.local.i` and `pop.local.i` to be completed with the rest of the local memory handling in stage 3 of #84
- updates and adds unit tests for all of the above

bobbinth commented on assembly/src/parsers/io_ops.rs:
```diff
@@ -207,32 +263,84 @@ pub fn parse_storew(span_ops: &mut Vec<Operation>, op: &Token) -> Result<(), Ass
+    param_idx: usize,
+    param_str: &str,
+) -> Result<Vec<BaseElement>, AssemblyError> {
+    let is_single_element = param_str.len() <= HEX_CHUNK_SIZE;
+    let is_invalid_param =
+        param_str.len() % 2 != 0 || !is_single_element && param_str.len() % HEX_CHUNK_SIZE != 0;
+
+    if is_invalid_param {
+        Err(AssemblyError::invalid_param(op, param_idx))
+    } else if param_str.len() > HEX_CHUNK_SIZE * MAX_CONST_INPUTS {
+        Err(AssemblyError::extra_param(op))
```
Took me a while to parse through this. Maybe we should improve readability here a bit (even at the expense of a few extra lines)? Maybe something like:
```Rust
let is_single_element = if param_str.len() <= HEX_CHUNK_SIZE {
    if param_str.len() % 2 != 0 {
        return Err(AssemblyError::invalid_param(op, param_idx));
    }
    true
} else {
    if param_str.len() % HEX_CHUNK_SIZE != 0 {
        return Err(AssemblyError::invalid_param(op, param_idx));
    } else if param_str.len() > HEX_CHUNK_SIZE * MAX_CONST_INPUTS {
        return Err(AssemblyError::extra_param(op));
    }
    false
};
```
And then we can handle non-error cases after this.

grjte replied:
Yeah, this makes sense.

bobbinth reviewed (changes requested):
Looks great! Thank you! I left just one comment related to readability inline.

Another thing I'm wondering: since the file is getting rather big, should we split out tests into a separate file? Basically, replace `io_ops.rs` with `io_ops` directory containing `mod.rs` and `tests.rs` files.

grjte commented:
> Another thing I'm wondering: since the file is getting rather big, should we split out tests into a separate file? 

I see your point, and I don't feel very strongly, but personally I would lean towards keeping it as it is. The idiomatic structure for Rust is to keep unit tests in the same file and have separate test files be for integration tests, and I like the current organization of the parsers. The addition of the locals handling + tests will increase it by a couple hundred lines at most.

One thing that I've thought about which could address this is that while the "invalid" tests that ensure ops are well-formed are very useful and do belong here, I'm not as convinced by the "valid" tests like ```popw_mem``` or ```push_one```. I wrote them on the first implementation to get my head around what was going on, but they feel sort of redundant and ineffective. Maybe this checking should be in the processor integration tests instead and removed from here. That would considerably shorten the file length

bobbinth reviewed (approved):
Looks great!

Regarding keeping tests in the same file: I think it is fine to do it that way if that's your preference. I tend to split them out into separate file once files get over 500 - 600 lines (while still keeping the same module structure) - makes it a bit easier for me, even though it is not the recommended way. But either way is fine.

jonathanxuu commented on assembly/src/parsers/io_ops.rs:
```diff
@@ -198,7 +254,7 @@ pub fn parse_storew(span_ops: &mut Vec<Operation>, op: &Token) -> Result<(), Ass
 
     match op.parts()[1] {
-        "mem" => parse_mem_write(span_ops, op),
+        "mem" => parse_mem_write(span_ops, op, true),
```
> The memory address should be removed from the stack by `STOREW` according to [spec](https://hackmd.io/YDbjUVHTRn64F4LPelC-NA?view#Random-access-memory)


Perhaps, the preserve_stack should be `false` here.

grjte commented:
I think it's correct, but perhaps the variable is poorly named, and maybe it could also use inline comments to make it clearer. What do you think?

The idea is that with ```storew``` the top 4 elements are written to mem and then left on the stack (the stack is preserved), whereas with ```popw``` they are written to mem and then popped from the stack (the stack is not preserved). The address is always removed from the stack during all mem ops, so the "preserve_stack" bool is only talking about the rest of the stack, not about the address.

------- Original Message -------
On Friday, February 11th, 2022 at 6:40 AM, jonathanxuu ***@***.***> wrote:

> @jonathanxuu commented on this pull request.
>
> ---------------------------------------------------------------
>
> In [assembly/src/parsers/io_ops.rs](https://github.com/maticnetwork/miden/pull/104#discussion_r804388429):
>
>> @@ -198,7 +254,7 @@ pub fn parse_storew(span_ops: &mut Vec<Operation>, op: &Token) -> Result<(), Ass
>      }
>
>      match op.parts()[1] {
> -        "mem" => parse_mem_write(span_ops, op),
> +        "mem" => parse_mem_write(span_ops, op, true),
>
>> The memory address should be removed from the stack by STOREW according to [spec](https://hackmd.io/YDbjUVHTRn64F4LPelC-NA?view#Random-access-memory)
>
> Perhaps, the preserve_stack should be false here.
>
> —
> Reply to this email directly, [view it on GitHub](https://github.com/maticnetwork/miden/pull/104#pullrequestreview-879815069), or [unsubscribe](https://github.com/notifications/unsubscribe-auth/AV2CLYJLKYQ7HYUDV7QGQNLU2SVPVANCNFSM5N43777Q).
> Triage notifications on the go with GitHub Mobile for [iOS](https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675) or [Android](https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub).
> You are receiving this because you authored the thread.Message ID: ***@***.***>

jonathanxuu commented:
这是来自QQ邮箱的自动回复邮件。徐毅正已经收到了您的邮件。
